### PR TITLE
#164: Added validation check before writing to intermediate location

### LINKED ISSUE #164: Validate that the write directory does not exist
From discussion https://github.com/exasol/spark-connector/pull/161#discussion_r1205117959

### PR DESCRIPTION
**Pull request by morazow**
Fixes #164

### CODE REVIEW

File: doc/development/developer_guide.md [hunk 1]
```diff
@@ -1,23 +1,28 @@
...
+
+The path layout:
 
-The integration tests are run using [Docker][docker] containers. The tests use
-[exasol-testcontainers][exa-testcontainers] and
-[spark-testing-base][spark-testing-base].
+```
+userProvidedS3Bucket/
+└── uuid-sparkApplicationId
+    └── queryId
```
Comment (kaklakariada) [inline doc/development/developer_guide.md:12]:
```suggestion
userProvidedS3Bucket/
└── <UUID>-<SparkApplicationId>/
    └── <SparkQueryId>/
```

File: doc/development/developer_guide.md [hunk 2]
```diff
@@ -1,23 +1,28 @@
...
+```
+
+The generated intermediate write path `<UUID>-<SparkApplicationId>/<SparkQueryId>/` if validated that it is empty before write. And it is cleaned up after the write query finishes.
```
Comment (kaklakariada) [inline doc/development/developer_guide.md:15]:
```suggestion
The generated intermediate write path `<UUID>-<SparkApplicationId>/<SparkQueryId>/` is validated that it is empty before write. And it is cleaned up after the write query finishes.
```

File: doc/development/developer_guide.md [hunk 3]
```diff
@@ -1,23 +1,28 @@
...
-[spark-testing-base]: https://github.com/holdenk/spark-testing-base
-[dev-guide]: https://github.com/exasol/import-export-udf-common-scala/blob/master/doc/development/developer_guide.md
+The integration tests are run using [Docker](https://www.docker.com) and [exasol-testcontainers](https://github.com/exasol/exasol-testcontainers/)
 
 ## Release
 
 Currently [release-droid](https://github.com/exasol/release-droid) has some troubles releasing the spark connector.
 
-* **Validation**: When using the local folder with `release-droid -l .` then validation failes with error message `NumberFormatException for input string: "v0"`.<br />
+* **Validation**: When using the local folder with `release-droid -l .` then validation fails with error message `NumberFormatException for input string: "v0"`.<br />
```
Comment (kaklakariada) [inline doc/development/developer_guide.md:25]: This was fixed in release droid, please remove the note.

File: exasol-s3/src/main/java/com/exasol/spark/s3/ExasolWriteBuilderProvider.java [hunk 4]
```diff
@@ -44,7 +48,23 @@ public ExasolWriteBuilderProvider(final ExasolOptions options) {
...
+        final String s3BucketKey = this.s3BucketKeyPathProvider.getS3BucketKeyForWriteLocation(defaultInfo.queryId());
+        validateWritePathIsEmpty(s3Bucket, s3BucketKey);
+        final LogicalWriteInfo info = getUpdatedLogicalWriteInfo(defaultInfo, s3Bucket, s3BucketKey, sparkSession);
+        return createCSVWriteBuilder(sparkSession, schema, info);
+    }
+
+    private void validateWritePathIsEmpty(final String s3Bucket, final String s3BucketKey) {
+        try (final S3FileSystem s3FileSystem = S3FileSystem.fromOptions(this.options)) {
+            if (!s3FileSystem.isEmpty(s3Bucket, Optional.of(s3BucketKey))) {
+                throw new ExasolValidationException("bucketIs not empty");
```
Comment (kaklakariada) [inline exasol-s3/src/main/java/com/exasol/spark/s3/ExasolWriteBuilderProvider.java:61]: Use error reporting if possible and add bucket name and key to the error message

File: exasol-s3/src/test/java/com/exasol/spark/s3/ExasolWriteBuilderProviderIT.java [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+                s3BucketKeyPathProvider);
+        assertDoesNotThrow(() -> writeBuilderProvider.createWriteBuilder(schema, getLogicalWriteInfo("queryId1")));
+    }
+
+    @Test
+    void testThrowsValidateWritePathWhenNotEmpty() throws IOException {
+        final String bucketKeyPrefix = s3BucketKey + "-" + applicationId + "/queryIdTest";
+        PutObjectRequest objectRequest = PutObjectRequest.builder().bucket(DEFAULT_BUCKET_NAME)
+                .key(bucketKeyPrefix + "/hello.txt").build();
+        s3Client.putObject(objectRequest, RequestBody.fromByteBuffer(getRandomByteBuffer(10_000)));
```
Comment (kaklakariada) [inline exasol-s3/src/test/java/com/exasol/spark/s3/ExasolWriteBuilderProviderIT.java:58]:
```suggestion
        s3Client.putObject(objectRequest, RequestBody.fromByteBuffer(getRandomByteBuffer(10)));
```
The file size should not matter, we only want to check if it exists

**Review by kaklakariada — CHANGES_REQUESTED** [review]

File: doc/development/developer_guide.md (hunk 3 above)
Comment (morazow) [reply doc/development/developer_guide.md:25]: Ahh great, I also had similar impression, but was not exactly sure.

File: exasol-s3/src/main/java/com/exasol/spark/s3/ExasolWriteBuilderProvider.java (hunk 4 above)
Comment (morazow) [reply exasol-s3/src/main/java/com/exasol/spark/s3/ExasolWriteBuilderProvider.java:61]: Opps, forgot to update this. Going to add error, message and code.

File: doc/development/developer_guide.md (hunk 3 above)
Comment (morazow) [reply doc/development/developer_guide.md]: Removed whole section, since `relese_config.yml` file already contains language.

**Review by kaklakariada — APPROVED** [review]

**Pull request merged**